src/org/rascalmpl/library/lang/rascal/tests/concrete/recovery/ParseErrorDisambiguator.java

```diff
@@ -0,0 +1,174 @@
+
+public class ParseErrorDisambiguator {
+	private final IRascalValueFactory rascalValues;
+    private boolean allowAmbiguity;
+
+    public ParseErrorDisambiguator(IRascalValueFactory rascalValues) {
+		super();
+		
+		this.rascalValues = rascalValues;
+    }
+
+    private static class ScoredTree {
+        public final IConstructor tree;
+        public final int score;
+
+        public ScoredTree(IConstructor tree, int score) {
+            this.tree = tree;
+            this.score = score;
+        }
+    }
+
+    /**
+     * Disambiguate error trees, faster implementation of the original Rascal variant:
+Tree defaultErrorDisambiguationFilter(Tree t) {
+  return visit(t) {
+    case a:amb(_) => ambDisambiguation(a)
+  };
+}
+
+private Tree ambDisambiguation(amb(set[Tree] alternatives)) {
+  // Go depth-first
+  rel[int score, Tree alt] scoredErrorTrees = { <scoreErrors(alt), alt> | Tree alt <- alternatives };
+  set[Tree] nonErrorTrees = scoredErrorTrees[0];
+
+  if (nonErrorTrees == {}) {
+    return (getFirstFrom(scoredErrorTrees) | it.score > c.score ? c : it | c <- scoredErrorTrees).alt;
+  }
+  
+  if ({Tree single} := nonErrorTrees) {
+    // One ambiguity left, no ambiguity concerns here
+    return single;
+  }
+
+  // Multiple non-error trees left, return an ambiguity node with just the non-error trees
+  return amb(nonErrorTrees);
+}
+
+private int scoreErrors(Tree t) = (0 | it + getSkipped(e).src.length | /e:appl(error(_,_,_),_) := t);
+
+// Handle char and cycle nodes
+default Tree defaultErrorDisambiguationFilter(Tree t) = t;
+     **/
+
+    public synchronized IConstructor disambiguateErrors(IConstructor arg, IBool allowAmbiguity) {
+        this.allowAmbiguity = allowAmbiguity != null && allowAmbiguity.getValue();
+        return disambiguate(arg).tree;
+    }
+
+    private ScoredTree disambiguate(IConstructor tree) {
+        Type type = tree.getConstructorType();
+
+		if (type == RascalValueFactory.Tree_Appl) {
+            return disambiguateAppl((ITree) tree);
+        } else if (type == RascalValueFactory.Tree_Amb) {
+            return disambiguateAmb((ITree) tree);
+        }
+
+        return new ScoredTree(tree, 0);
+    }
+
+    private ScoredTree disambiguateAppl(ITree appl) {
+        if (ProductionAdapter.isSkipped(appl.getProduction())) {
+            return new ScoredTree(appl, ((IList) appl.get(1)).length());
+        }
+
+        IList args = TreeAdapter.getArgs(appl);
+        int totalScore = 0;
+        IValue[] disambiguatedArgs = new IValue[args.size()];
+
+        int index=0;
+        boolean anyChanges = false;
+
+        // Disambiguate and score all children
+        for (IValue arg : args) {
+            ScoredTree disambiguatedArg = disambiguate((IConstructor) arg);
+            totalScore += disambiguatedArg.score;
+            disambiguatedArgs[index++] = disambiguatedArg.tree;
+            if (disambiguatedArg.tree != arg) {
+                anyChanges = true;
```

Comment:
we could also only allocate the `new IValue[]` after the first time a different tree came back. since otherwise we're building & allocating arrays all the time without use. and mayby in that case just allocate a list writer to begin with?

Reply:
I considered this but as this code is only used for `amb` nodes I decided against complicating the code until profiling indicates this is an issue.